### PR TITLE
Fixes #210: --output option doesn't work when input is stdin

### LINKED ISSUE #210: The output option doesn't seem to work.
As far as I can tell, `rnp` completely ignores the `--output` option, which makes tests more complicated because it's necessary to have multiple copies of input files in order to get different output files.

### PR DESCRIPTION
**Pull request by ni4**
For the rnp utility, during the signing/encryption/whatever else, --output option worked only for file input, i.e. when input filename is specified, and was ignored when input is stdin. Now it should work for stdin as well.

### CODE REVIEW

**Comment by catap** [comment]
@ni4 why have you merge it?

I think it's bad idea because I see at master right for example `include "../common/constants.h"` [here](https://github.com/riboseinc/rnp/blob/master/include/rnp.h#L36)

and this is huge PR and I think it should be reviewed more of one person, and review this huge changes request a lot of time.

File: src/rnp/rnp.c [hunk 1]
```diff
@@ -221,24 +223,55 @@ stdin_to_mem(rnp_cfg_t *cfg, char **temp, char **out, unsigned *maxsize)
...
+            flags |= O_TRUNC;
+        else
+            flags |= O_EXCL;
+
+        fd = open(outfile, flags, 0600);
+        if (fd < 0) {
+            if (overwrite && (errno == EEXIST)) {
+                fprintf(stderr, "Failed to write to the %s : file already exists.\n", outfile);
+            } else {
+                fprintf(
```
Comment (catap) [inline src/rnp/rnp.c]: I think here better to use `strerror` or `perror`

File: src/rnp/rnp.c [hunk 2]
```diff
@@ -221,24 +223,55 @@ stdin_to_mem(rnp_cfg_t *cfg, char **temp, char **out, unsigned *maxsize)
...
+                  stderr, "Failed to open file %s for writing : error %d.\n", outfile, errno);
+            }
+            return RNP_FAIL;
+        }
+    }
+
     for (cc = 0; cc < size; cc += n) {
-        if ((n = write(STDOUT_FILENO, &out[cc], size - cc)) <= 0) {
+        if ((n = write(fd, &out[cc], size - cc)) <= 0) {
```
Comment (catap) [inline src/rnp/rnp.c]: maybe also show the error?

**Comment by catap** [comment]
1) I think that better keep different changes as different commit. Here I see changes inside two different files and I have no idea why you change `src/lib/rnp.c`.

2) I have no idea that the commit `--output option doesn't work when input is stdin` means. This just broke me :(

File: src/lib/rnp.c [hunk 3]
```diff
@@ -1072,24 +1072,23 @@ rnp_encrypt_file(rnp_ctx_t *ctx, const char *userid, const char *f, const char *
...
         return RNP_FAIL;
     }
-    suffix = (ctx->armour) ? ".asc" : ".gpg";
     /* get key with which to sign */
     if ((key = resolve_userid(ctx->rnp, ctx->rnp->pubring, userid)) == NULL) {
         return RNP_FAIL;
     }
+    /* generate output file name if needed */
     if (out == NULL) {
+        suffix = (ctx->armour) ? ".asc" : ".gpg";
```
Comment (ronaldtse) [inline src/lib/rnp.c]: I think we need a better method than hard-coding file extensions? I know it's a moved line but maybe for the future.

File: src/rnp/rnp.c (hunk 1 above)
Comment (ni4) [inline src/rnp/rnp.c]: Thanks, will fix this.

File: src/rnp/rnp.c (hunk 2 above)
Comment (ni4) [inline src/rnp/rnp.c]: Ok, will improve this.

File: src/lib/rnp.c (hunk 3 above)
Comment (ni4) [inline src/lib/rnp.c]: Yeah, I don't like this one as well. Just didn't fix this since it went a bit far from the purpose of this commit.

**Comment by ni4** [comment]
@catap , thanks for the review.
1.  It was just some cleanup in lib/rnp.c which I forgot to do in previous PR
2. I updated the description, also commit includes link to the issue #210 where it was discussed.

**Comment by catap** [comment]
@ni4, ok, can you also update the commit log by `git rebase -i`? And make style-checker happy!

**Comment by dewyatt** [comment]
FYI I haven't really tested the pre-commit hook with rebasing/amending much. And thinking about it now, it probably *will* do things you don't want (though you shouldn't lose data). Anyways, it may be better to run clang-format manually and then squash (and then enable the pre-commit hook).

**Comment by ni4** [comment]
Sure, I'll update this commit as soon as I get to it. A bit busy during these days. Usually I run clang-format manually before each push to PR, but this time I forgot.

**Comment by ni4** [comment]
@ronaldtse @catap Please see the updated PR. Also, since you mentioned rebase. Do we merge or rebase? This should be also described in developers guide, I think.

**Comment by catap** [comment]
@ni4 I do rebase, but merge it's easy :) So, I think it is up to you in general.

But rebase helps commits log clear and smaller.

**Comment by dewyatt** [comment]
Agree with @catap that it's something to evaluate for every PR, not something to set a solid rule on.
Ronald usually does a plain merge, I usually rebase, but sometimes squash when there are a lot of minor "fix it" commits.

**Comment by ronaldtse** [comment]
@dewyatt personally I prefer a rebase, but a merge is good to preserve the commits (given they are clean) for an easier revert. Now that this is mentioned I'll stick to a rebase 👍

**Comment by ni4** [comment]
Thanks for the comments. I would appreciate some approvals or requests for changes so I can rebase and merge this :)

**Comment by dewyatt** [comment]
> Thanks for the comments. I would appreciate some approvals or requests for changes so I can rebase and merge this :)

LGTM (but note that the reviewer should merge, not the PR creator)

File: src/rnp/rnp.c (hunk 1 above)
Comment (ronaldtse) [inline src/rnp/rnp.c]: @ni4 is this fixed? Thanks.

File: src/rnp/rnp.c (hunk 2 above)
Comment (ronaldtse) [inline src/rnp/rnp.c]: @ni4 just following up.

File: src/rnp/rnp.c [hunk 4]
```diff
@@ -315,10 +347,12 @@ rnp_cmd(rnp_cfg_t *cfg, rnp_t *rnp, int cmd, char *f)
         ctx.sigcreate = get_birthtime(rnp_cfg_get(cfg, CFG_BIRTHTIME));
         ctx.sigexpire = get_duration(rnp_cfg_get(cfg, CFG_DURATION));
 
+        clearsign = (cmd == CMD_CLEARSIGN) ? true : false;
```
Comment (flowher) [inline src/rnp/rnp.c]: I think you should also declare ``clearsign`` as ``bool``. Also 
``clearsign = (cmd == CMD_CLEARSIGN)``
means same thing with less code

Comment (ni4) [inline src/rnp/rnp.c]: @flowher I believe I changed it to bool (among with other things in rnp.c/rnpcfg.c) in some other branch which did not become pull request yet :) Sorry, was a bit out of rnp this week, will get back tomorrow.
@ronaldtse everything should be fixed here, just was waiting for approvals.

**Comment by ronaldtse** [comment]
I'm just going to merge this first and leave the comments as separate issues.

**Comment by ronaldtse** [comment]
Thanks @ni4 !

**Pull request merged**